datafusion/src/error.rs

```diff
@@ -101,6 +104,12 @@ impl From<ParserError> for DataFusionError {
     }
 }
 
+impl From<Box<dyn Error + Send + Sync>> for DataFusionError {
+    fn from(err: Box<dyn Error + Send + Sync>) -> Self {
+        DataFusionError::External(err)
+    }
+}
+
```

Comment:
@houqp apologies if im misunderstanding, but is this what youre referring to?  Since i cant match on trait object i thought i was limited on the types of `DataFusionError` I could return with this.  I was definitely trying to make changes that would "bubble up" automatically but i must have done something incorrectly as thats clearly not happening.

Reply:
I was referring to something like this `let mut reader = obj_reader?.sync_reader().map_err(DatafusionError::from)?;` so the low level functions that call these object store apis can hide the Error trait objects from their callers.

Reply:
Ah - thank you for the clarification. I can try that. So all of these sites would now have External errors. 

I'm wondering if there would be value in having a macro or function that could downcast the trait object to concrete datafusionerrors (kind of like what we do with compute kernels to downcast to concrete array types)  - so we could really use external for true external code.

Maybe that's a next step though. Just thinking out loud.

Reply:
I should probably have a IOError or ObjectStoreError enum variant specifically for wrapping object store errors :)

As for downcasting error trait objects, it's best to be managed by the users because datafusion core will not have access to all different types of error that plugin object store will return. We can do that for arrow compute kernels because arrow is a dependency of datafusion.

Reply:
Sounds good and makes sense, thanks much @houqp.  I will work on that.

FYI there is existing `IoError(io::Error)` variant.  I think ill have to see how well that aligns with the types of errors we get in the `ObjectStore` (i.e are the `io::Error`?)